File: LifestyleFY/backend/app/services/categories.py
"""Fixed pantry category taxonomy, shared conceptually with the frontend's
core/categories.ts (hand-synced, same pattern as the Pydantic/TS model split).
"""
from __future__ import annotations

APP_CATEGORIES = [
    {"id": "produce", "label": "Produce", "emoji": "🥦"},
    {"id": "dairy", "label": "Dairy", "emoji": "🥛"},
    {"id": "meat", "label": "Meat & Seafood", "emoji": "🍗"},
    {"id": "frozen", "label": "Frozen", "emoji": "🧊"},
    {"id": "grains", "label": "Grains & Bread", "emoji": "🍞"},
    {"id": "snacks", "label": "Snacks", "emoji": "🍿"},
    {"id": "sweets", "label": "Sweets & Desserts", "emoji": "🍫"},
    {"id": "beverages", "label": "Beverages", "emoji": "🥤"},
    {"id": "condiments", "label": "Condiments & Sauces", "emoji": "🧂"},
    {"id": "canned", "label": "Canned & Jarred", "emoji": "🥫"},
    {"id": "other", "label": "Other", "emoji": "🧺"},
]

_DEFAULT_LOCATION = {
    "produce": "fridge",
    "dairy": "fridge",
    "meat": "fridge",
    "frozen": "freezer",
    "grains": "pantry",
    "snacks": "pantry",
    "sweets": "pantry",
    "beverages": "pantry",
    "condiments": "pantry",
    "canned": "pantry",
    "other": "pantry",
}

_PNNS_GROUP_1_MAP = {
    "Fruits and vegetables": "produce",
    "Milk and dairy products": "dairy",
    "Fish Meat Eggs": "meat",
    "Cereals and potatoes": "grains",
    "Salty snacks": "snacks",
    "Sugary snacks": "sweets",
    "Beverages": "beverages",
    "Fat and sauces": "condiments",
}


def category_from_off(pnns_group_1: str | None, categories_tags: list[str] | None) -> str:
    """Map OFF's pnns_groups_1 (+ a frozen-foods override) to an app category id.
    Always returns a valid id, defaulting to "other" when nothing matches."""
    tags = categories_tags or []
    if "en:frozen-foods" in tags:
        return "frozen"
    return _PNNS_GROUP_1_MAP.get(pnns_group_1 or "", "other")


def default_location(category: str | None) -> str:
    return _DEFAULT_LOCATION.get(category or "", "pantry")

File: LifestyleFY/frontend/src/app/core/categories.ts
// Mirrors the backend pantry category taxonomy (app/services/categories.py).

export type Location = 'pantry' | 'fridge' | 'freezer';

export interface CategoryMeta {
  id: string;
  label: string;
  emoji: string;
  defaultLocation: Location;
}

export const APP_CATEGORIES: CategoryMeta[] = [
  { id: 'produce', label: 'Produce', emoji: '🥦', defaultLocation: 'fridge' },
  { id: 'dairy', label: 'Dairy', emoji: '🥛', defaultLocation: 'fridge' },
  { id: 'meat', label: 'Meat & Seafood', emoji: '🍗', defaultLocation: 'fridge' },
  { id: 'frozen', label: 'Frozen', emoji: '🧊', defaultLocation: 'freezer' },
  { id: 'grains', label: 'Grains & Bread', emoji: '🍞', defaultLocation: 'pantry' },
  { id: 'snacks', label: 'Snacks', emoji: '🍿', defaultLocation: 'pantry' },
  { id: 'sweets', label: 'Sweets & Desserts', emoji: '🍫', defaultLocation: 'pantry' },
  { id: 'beverages', label: 'Beverages', emoji: '🥤', defaultLocation: 'pantry' },
  { id: 'condiments', label: 'Condiments & Sauces', emoji: '🧂', defaultLocation: 'pantry' },
  { id: 'canned', label: 'Canned & Jarred', emoji: '🥫', defaultLocation: 'pantry' },
  { id: 'other', label: 'Other', emoji: '🧺', defaultLocation: 'pantry' },
];

const OTHER = APP_CATEGORIES[APP_CATEGORIES.length - 1];
const BY_ID = new Map(APP_CATEGORIES.map((c) => [c.id, c]));

export function categoryMeta(id: string | null | undefined): CategoryMeta {
  return (id && BY_ID.get(id)) || OTHER;
}

export function defaultLocation(categoryId: string | null | undefined): Location {
  return categoryMeta(categoryId).defaultLocation;
}

export const LOCATIONS: { id: Location; label: string }[] = [
  { id: 'pantry', label: 'Pantry' },
  { id: 'fridge', label: 'Fridge' },
  { id: 'freezer', label: 'Freezer' },
];
